### DISCUSSION TITLE
does `url` in `NextRequest` always return a fully qualified URL in Middleware?

### DISCUSSION THREAD

**Discussion by blakewilson** [vercel/next.js#35151] [Help] [answered]
When using Next.js middleware, either on Vercel, custom server, or another hosting provider, will the `req.url` always be a full URL, with protocol and host? For example, in Middleware:

```tsx
export default async function middleware(req: NextRequest) {
  console.log('url', req.url); // -> http://localhost:3000/about?

  return NextResponse.next();
}
```

The reason I ask is because I have seen some instances when only the pathname is returned. For example, when using `getServerSideProps`:

```tsx
export async function getServerSideProps(context: GetServerSidePropsContext) {
  console.log(context?.req?.url); // -> /about

  return {
    props: {},
  };
}
```

EDIT:

In reference to https://github.com/vercel/next.js/pull/33410, it looks like middleware will always have an absolute URL in the Edge Functions runtime. In this case, is the Edge Functions runtime only applicable to Vercel, or does this runtime also operate when using a different provider or custom server? 

CC @javivelasco 

Thanks!

**Comment by javivelasco** [comment, marked as answer]
Yes! Next.JS Middleware works always with absolute URLs so you will always get a fully qualified URL. This is true for both production and development and for Vercel.
